plutus-chain-index-core/src/Plutus/ChainIndex/Handlers.hs

```diff
@@ -253,15 +259,16 @@ appendBlocks blocks = do
                 Right InsertUtxoSuccess{newIndex, insertPosition} -> do
                     logDebug $ InsertionSuccess tip_ insertPosition
                     return (newIndex, transactions ++ txs, newUtxoState : utxoStates)
-    oldIndex <- get @ChainIndexState
+    indexStateVar <- ask @(MVar ChainIndexState)
+    oldIndex <- liftIO $ takeMVar indexStateVar
```

Comment:
Ideally we should use `modifyMVar_` here to deal with exceptions but calling `Effs` in the `IO` is unpleasant.

I hope that `reduceBlockCount` doesn't throw implicit exceptions. Or we can replace `takeMVar` with `readMVar` and use swap later instead of `put`.